core/test/utils/matrix_utils_test.cpp

```diff
@@ -198,4 +201,77 @@ TYPED_TEST(MatrixUtils, MakeHpdMatrixWithRatioCorrectly)
 }
 
 
+TEST(MatrixUtils, RemoveDiagonalEntry)
+{
+    using T = float;
+    using Csr = gko::matrix::Csr<T, int>;
+    auto exec = gko::ReferenceExecutor::create();
+    auto b = gko::initialize<Csr>(
+        {I<T>{2.0, 0.0, 1.1, 0.0}, I<T>{1.0, 2.4, 0.0, -1.0},
+         I<T>{0.0, -4.0, 2.2, -2.0}, I<T>{0.0, -3.0, 1.5, 1.0}},
+        exec);
+    const int row_to_remove = 2;
+
+    gko::test::remove_diagonal_entry_from_row(b.get(), row_to_remove);
+
+    const auto rowptrs = b->get_const_row_ptrs();
+    const auto colidxs = b->get_const_col_idxs();
+    bool all_diags = true;
+    for (int i = 0; i < 3; i++) {
+        if (i == row_to_remove) {
+            continue;
+        }
+        bool has_diag = false;
+        for (int j = rowptrs[i]; j < rowptrs[i + 1]; j++) {
+            if (colidxs[j] == i) {
+                has_diag = true;
+            }
+        }
+        if (!has_diag) {
+            all_diags = false;
+            break;
+        }
+    }
+    ASSERT_TRUE(all_diags);
+    bool has_diag = false;
+    for (int j = rowptrs[row_to_remove]; j < rowptrs[row_to_remove + 1]; j++) {
+        if (colidxs[j] == row_to_remove) {
+            has_diag = true;
+        }
+    }
+    ASSERT_FALSE(has_diag);
```

Comment:
nit: this can be simplified a bit
```suggestion
    for (int i = 0; i < 3; i++) {
        bool has_diag = false;
        for (int j = rowptrs[i]; j < rowptrs[i + 1]; j++) {
            if (colidxs[j] == i) {
                has_diag = true;
            }
        }
        ASSERT_EQ(has_diag, i == row_to_remove);
    }
```
if you like std:: algorithms, you can also replace the inner loop by
```cpp
col_begin = colidxs + rowptrs[i];
col_end = colidxs + rowptrs[i + 1]
bool has_diag = std::find(col_begin, col_end, i) != col_end;
```